### PR TITLE
Update osimC3D and test

### PR DESCRIPTION
**Pull request by jimmyDunne**
#### osimC3D.m 
expose convertMillimeters2Meters(). This was previously performed internally only when writing the forces to file during and should be user-controlled. 

#### testOsimC3D.m 
Updated test so that it runs in significantly less time (120 seconds -> 57 seconds on my local machine). 
- I compare OpenSim table values, rather than Matlab Structures, since the conversion of tables to structs is expensive. 
- I only test rotations about 1 axis. 
- Added test for convertMillimeters2Meters() (not previously tested).

#### c3dExport.m 
included improved use of convertMillimeters2Meters() in example script

### Testing I've completed
Running ctest locally (passed) and checking the outputs on various C3D files (locally). 

### CHANGELOG.md (choose one)
- updated since users will have to change their code to explicitly use convertMillimeters2Meters() before writing to file.

---
This change is [Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/2556)

### CODE REVIEW

File: Bindings/Java/Matlab/Utilities/osimC3D.m [hunk 1]
```diff
@@ -150,6 +150,32 @@ function rotateData(obj,axis,value)
             obj.rotateTable(obj.forces, axis, value);
             disp('Marker and Force tables have been rotated')
         end
+        function convertMillimeters2Meters(obj)
```
Comment (aymanhab) [inline Bindings/Java/Matlab/Utilities/osimC3D.m]: Can we rename obj to osimC3Dobject for clarity?

File: Bindings/Java/Matlab/Utilities/osimC3D.m [hunk 2]
```diff
@@ -150,6 +150,32 @@ function rotateData(obj,axis,value)
...
+            import org.opensim.modeling.*
+            
+            nRows  = obj.forces.getNumRows();
+            labels = obj.forces.getColumnLabels();
+            
+            for i = 0 : obj.forces.getNumColumns - 1
+                % All force columns will have the 'f' prefix while point
+                % and moment columns will have 'p' and 'm' prefixes,
+                % respectively. 
+                if ~contains(char(labels.get(i)),'f')
```
Comment (aymanhab) [inline Bindings/Java/Matlab/Utilities/osimC3D.m]: Maybe unrelated to this PR but is the 'f' prefix an arbitrary convention we made or is it enforced by C3D?

File: Bindings/Java/Matlab/Utilities/osimC3D.m (hunk 1 above)
Comment (jimmyDunne) [reply Bindings/Java/Matlab/Utilities/osimC3D.m]: As far as I am aware, it can only be obj (obj is itself).

File: Bindings/Java/Matlab/Utilities/osimC3D.m (hunk 2 above)
Comment (jimmyDunne) [reply Bindings/Java/Matlab/Utilities/osimC3D.m]: Yep, forces are labelled _fn_, where n is the force number. We use _p_ for COP and _m_ for Moment

File: Bindings/Java/Matlab/Utilities/osimC3D.m (hunk 1 above)
Comment (aymanhab) [reply Bindings/Java/Matlab/Utilities/osimC3D.m]: So on line 40 we can't name the object as c3dObject for example?? I guess then users are accustomed to confusion! My bad

Comment (jimmyDunne) [reply Bindings/Java/Matlab/Utilities/osimC3D.m]: I checked out some documentation from a few different sources, and you are correct, I can change this into a custom name. However, all examples I had ever seen used _obj_ , with some of the examples I found today using _self_. So I am hesitant to change it since someone new to OOP in Matlab may nor realize the mapping.

File: Bindings/Java/Matlab/Utilities/osimC3D.m [hunk 3]
```diff
@@ -150,6 +150,32 @@ function rotateData(obj,axis,value)
...
+                    for u = 0 : nRows - 1
+                        % Get the Vec3
+                        vec3 = obj.forces.getDependentColumnAtIndex(i).get(u);
+                        % Divide the Vec3 by 1000
+                        vec3_new = Vec3(vec3.get(0)/1000,vec3.get(1)/1000,vec3.get(2)/1000);
+                        % set the table value
```
Comment (aymanhab) [inline Bindings/Java/Matlab/Utilities/osimC3D.m]: This is excessively expensive to divide a Vec3 by a double, I''m looking into exposing the equivalent of operator for +=,-=, *=, /= to Java/Matlab so avoid the inefficiency/overhead but looks fine considering what we have now.

File: Bindings/Java/Matlab/Utilities/osimC3D.m [hunk 4]
```diff
@@ -170,10 +196,6 @@ function writeMOT(obj,varargin)
         % osimC3d.writeMOT()                       Write to dir of input c3d.
         % osimC3d.writeMOT('Walking.mot')          Write to dir of input c3d with defined file name.
         % osimC3d.writeMOT('C:/data/Walking.mot')  Write to defined path input path.
-        %
```
Comment (aymanhab) [inline Bindings/Java/Matlab/Utilities/osimC3D.m]: Can we remove these comments unless you use for debugging, I was worried about sneaking absolute paths into user facing code.

File: Bindings/Java/Matlab/tests/testOsimC3D.m [hunk 5]
```diff
@@ -34,76 +34,92 @@
...
+randomRows = randsample(forces.getNumRows(),101)-1;
+for i =  0 : forces_m.getNumColumns() - 1
+    
+    col =  forces_rotated.getDependentColumnAtIndex(i);
+    col_m = forces_m.getDependentColumnAtIndex(i);
+    label = forces_m.getColumnLabel(i);
+    
+    for u = 1 : length(randomRows)
+        if contains(char(forces_m.getColumnLabels().get(i)),'f')
+            % Get the values
```
Comment (aymanhab) [inline Bindings/Java/Matlab/tests/testOsimC3D.m]: Should this at least be start with 'f' rather than contains 'f'? Not convinced that making up a convention while reading data is the right thing to do, but seems unrelated.

File: Bindings/Java/Matlab/Utilities/osimC3D.m (hunk 1 above)
Comment (aymanhab) [reply Bindings/Java/Matlab/Utilities/osimC3D.m]: Also can we verify the units instead of assuming them to avoid potentially corrupting data or are the units not built into the format?

Comment (aymanhab) [reply Bindings/Java/Matlab/Utilities/osimC3D.m]: I would prefer 'self' which means something special over a generic 'obj' that can easily get lost, but you're call. Thanks for investigating.

File: Bindings/Java/Matlab/Utilities/osimC3D.m (hunk 4 above)
Comment (jimmyDunne) [reply Bindings/Java/Matlab/Utilities/osimC3D.m]: They are actually help statements for the function that a user can call;

``` java
>> help osimC3D/writeMot
  Write force data to mot file.
  osimC3d.writeMOT()                       Write to dir of input c3d.
  osimC3d.writeMOT('Walking.mot')          Write to dir of input c3d with defined file name.
  osimC3d.writeMOT('C:/data/Walking.mot')  Write to defined path input path.
  
  This function assumes point and torque data are in mm and Nmm and
  converts them to m and Nm. If your C3D is already in M and Nm,
  comment out the internal function convertMillimeters2Meters()
```
In this case, you can use either of the three methods (empty, filename, or full path) to write the output.

File: Bindings/Java/Matlab/tests/testOsimC3D.m (hunk 5 above)
Comment (jimmyDunne) [reply Bindings/Java/Matlab/tests/testOsimC3D.m]: I preferred _contains()_  in this case because it will return a 1 or 0. If I use something like _strfind()_ it will return either 1 or [], which means I have to do something like ~isempty(strfind()).  Since we (_C3DFileAdapter()_) are the ones that define the column names, we are guaranteed of not having a case when there may be a rogue _f_ in the column name (like if someone named their column COP_f for some reason).

Comment (aymanhab) [reply Bindings/Java/Matlab/tests/testOsimC3D.m]: ok, if there's nothing like startsWith() that returns logical.

File: Bindings/Java/Matlab/Utilities/osimC3D.m (hunk 1 above)
Comment (jimmyDunne) [reply Bindings/Java/Matlab/Utilities/osimC3D.m]: I agree, _self_ feels like it would be a better differentiator than _obj_. I had always seen _obj_ in examples so went with that. I will update to _self_ here since it doesn't break anything (I tested it locally already).

Comment (jimmyDunne) [reply Bindings/Java/Matlab/Utilities/osimC3D.m]: > Also can we verify the units instead of assuming them to avoid potentially corrupting data or are the units not built into the format?

Agreed, but there isn't information on units in the Forces table metadata.  There is probably (almost certainly) a way of getting it from _C3DFileAdapter()_ and including it in the metadata.

The reasons for exposing _convertMillimeters2Meters()_ was actually because of this. I had been converting everyones data by default at write time, which is probably fine in most cases (Vicon stores in mm) but it could cause havoc on if not.

File: Bindings/Java/Matlab/tests/testOsimC3D.m (hunk 5 above)
Comment (jimmyDunne) [reply Bindings/Java/Matlab/tests/testOsimC3D.m]: Oh, SNAP! hahaha, I hadn't seen that! Thanks! Perfect I will change to startsWith()

File: Bindings/Java/Matlab/Utilities/osimC3D.m (hunk 3 above)
Comment (aymanhab) [reply Bindings/Java/Matlab/Utilities/osimC3D.m]: PR #2558 should allow you to replace this with `vec3.scalarDivideEq(1000)` in a fraction of the time but don't hold this PR off for it.

**Comment by jenhicks** [comment]
@aymanhab will take a pass through the latest. Need to update here and elsewhere (maybe) to leverage new methods for vector multiplication/division by a scalar.

**Comment by aymanhab** [comment]
Thanks @jimmyDunne I merged this and #2558, should be minor change to upgrade after. I didn't delete the branch in case you want to reuse it.
